keylime/da/attest.py

```diff
@@ -0,0 +1,154 @@
+    )
+
+    args = parser.parse_args()
+
+    rmcb = config.get("registrar", "durable_attestation_import", fallback="")
+    rmc = record.get_record_mgt_class(rmcb)
+    if not rmc:
+        return
+
+    rmc = rmc("registrar")
+
+    if args.end_date == "last":
+        end_date = rmc.end_of_times
+    else:
+        end_date = int(datetime.fromisoformat(args.end_date).strftime("%s"))
+
+    if args.start_date == "first":
+        start_date = rmc.start_of_times
+    elif args.start_date == "none":
+        start_date = end_date - 1
+    else:
+        start_date = int(datetime.fromisoformat(args.start_date).strftime("%s"))
+
+    if end_date != rmc.end_of_times or start_date != rmc.end_of_times - 1:
+        logger.info(
+            "=> Selected date range for attestation records is: start %d (%s) and end %d (%s)",
+            start_date,
+            datetime.fromtimestamp(start_date).strftime("%Y-%m-%dT%H:%M:%S.%f%z"),
+            end_date,
+            datetime.fromtimestamp(end_date).strftime("%Y-%m-%dT%H:%M:%S.%f%z"),
+        )
+    else:
+        logger.info("=> Will only extract the very last attestation record found on the persistent store")
+
+    if args.agent_uuid.lower() == "all":
+        logger.info("=> Getting a list of agents registered on the persistent store")
+        agent_list = rmc.agent_list_retrieval()
+    else:
+        logger.info("=> Focusing on the agent %s on the persistent store.", args.agent_uuid)
+        agent_list = [args.agent_uuid]
+
+    for agent_uuid in agent_list:
+        logger.info("===> Getting all existing registration records for agent %s...", agent_uuid)
+
+        ak_list = rmc.build_key_list(agent_uuid)
+
+        if not ak_list:
+            logger.error("Unable to assemble a list of AKs for the agent agent %s", agent_uuid)
+            return
+
+        logger.info("===> Getting all existing attestation records for agent %s ...", agent_uuid)
+        attestation_record_list = rmc.record_read(agent_uuid, start_date, end_date, "verifier")
+
+        logger.info("=====> Verifing the state of agent %s over time...", agent_uuid)
+
+        p_tpm_ts = 0
+        c_tpm_ts = 0
+        d_tpm_ts = 0
+
+        agentAttestState = None
+        for attestation_record in attestation_record_list:
+            agent = attestation_record["agent"]
+            json_response = attestation_record["json_response"]
+            ima_policy = attestation_record["ima_policy"]
+
+            logger.info(
+                "----------- Attesting data (quote and logs from %s, captured by verifier %s (%s:%s)",
+                agent_uuid,
+                agent["verifier_id"],
+                agent["verifier_ip"],
+                agent["verifier_port"],
+            )
+
+            if not agentAttestState:
+                agentAttestState = cloud_verifier_common.get_AgentAttestStates().get_by_agent_id(agent["agent_id"])
```

Comment:
Does this work as-is for "all" or do you not have to get the agentAttestState for the agent with the current agent's uuid? I assume the attestation_records are all coming in no specific order if "all" is requested.

Reply:
You need to get `agentAttestState` for current `uuid`, even in case of `all`.

Reply:
My point is you need an agentAttestState **per agent** given the agent's UUID.

Reply:
... the `if not agentAttestState` check should be removed, especially when 'all' is selected and you get different agents' measuements on every step in this loop

Reply:
@maugustosilva This seems like the last thing left unresolved.

Reply:
I believe there is a misunderstanding here... for **every agent**, `agentAttestState` is set to `None` and then it is initialized at the "beginning of time" for **that particular agent** (just once). The code basically **works as intended** with either a given agent UUID or "all", I am not really getting the need for any change here (I just pushed an unrelated change)

Reply:
At the moment there's no need since durable attestation replay with IMA doesn't work at all. However, the correct solution is to use an agentAttestState per agent whenever its PCRs need to be extended.

Reply:
Durable attestation with IMA works, **if** one is ok with replaying all attestation records, since the "beginning of time"

Reply:
It works only with a single machine otherwise you will have to change that line with the agentAttestState.

I think users should be made aware of these conditions.

Reply:
... and also it only works if one doesn't restart attestation altogether (after failure to attest or simple stop and update/resume of policy for example) because then you will have to reset the agentAttestState for that particular agent as well (so the virtual PCR 10 is reset) once you come across a boot_aggregate in the IMA measurement log.

Reply:
I did not test it with attestation restart (for instance, if the agent gets restarted, a new registration entry would be generated anyway), but again, instead of extensively listing all the cases where it would not work, I am explicitly stating this is an **experimental** feature. The **main goal** of this PR is to provide a way to **test the feature on the CI**.